src/pytest_alembic/tests/experimental/all_models_register_on_metadata.py

```diff
@@ -0,0 +1,191 @@
+    to import said model somewhere in the normal loading process for `foo.models`.
+
+    Thus this test compares such an import against the opposite scenario, one where
+    every single possible module has been imported, and we therefore "**know**" that
+    all defined models/tables will be attached to the MetaData. If there's a
+    difference in the set of tables on the MetaData, then we have an issue!
+
+    **Note** This test is distinct from the `test_model_definitions_match_ddl`
+    test because imports can have side-effects. Thus when normal tests run in the
+    context of your test run, you may have already imported models from modules
+    transitively somewhere else in tree of imports required to run your tests.
+
+    Therefore this test goes to lengths to collect a clean python interpreter
+    which **only** imports the portion of code which would have been imported by
+    the `env.py` itself.
+    """
+
+    modules, bare_tables = get_bare_import_tableset(
+        str(alembic_runner.connection_executor.connection.url),
+    )
+    full_tables = get_full_tableset(*modules)
+
+    if bare_tables != full_tables:
+        diff = bare_tables.symmetric_difference(full_tables)
+        diff_str = ", ".join(diff)
+        raise AlembicTestFailure(
+            "There was a difference detected between the set of tables registered on your "
+            "`MetaData` during the course of loading your `env.py` compared with the set of "
+            f"tables registered after importing all modules below the provided '{model_package}' "
+            "module/package.\n\n"
+            "This commonly happens if you are missing an import path from the root of your "
+            "models/MetaData to some set of models defined elsewhere. Given that your `env.py` "
+            "likely only directly imports your model base or MetaData, this can be a problem! "
+            "As far as alembic is concerned, affected tables do not exist, and will likely "
+            "be indicated by alembic to be dropped, should you attempt a `revision --autogenerate`"
+            "command.\n\n"
+            f"The following tables are affected: {diff_str}."
+        )
+
+
+def get_bare_import_tableset(url: str) -> Tuple[List[str], Set[str]]:
+    """Get the set of tables which would have been added to the metadata on a bare ma.models import.
+
+    Importantly, this cannot simply import the MetaData directly, as that may have already
+    had side-effects performed on it such that it's no longer representative of if one were to
+    directly import the MetaData alone.
+
+    Thus, we start a python subprocess which does the most minimal import we can perform to extract the MetaData,
+    and return the set of tables attached to it.
+    """
+
+    command = [
+        "python",
+        "-m",
+        "pytest_alembic.tests.experimental.collect_clean_alembic_environment",
+        shlex.quote(url),
```

Comment:
I don't think you want to quote the URL here: if you're not running with `shell=True` (which you shouldn't be, and aren't), the values of these strings show up verbatim in the subprocess' sys.argv. If this is working as-is, it's probably because none of the URLs you've passed actually contain any shell-special characters.

Reply:
good to know!